File: nestery-backend/src/integrations/oyo/oyo.service.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import axios from 'axios';
import { LoggerService } from '../../core/logger/logger.service';
import { ExceptionService } from '../../core/exception/exception.service';
import * as crypto from 'crypto';

// Define interfaces for OYO API responses
interface OyoHotelImage {
  url: string;
  caption?: string;
}

interface OyoHotelFacility {
  id: string;
  description: string;
  category?: string;
}

interface OyoHotelRoom {
  id: string;
  name: string;
  description: string;
  maxOccupancy: number;
  price: number;
  currency: string;
  facilities?: OyoHotelFacility[];
  images?: OyoHotelImage[];
}

interface OyoHotelReview {
  id: string;
  rating: number;
  comment: string;
  author: string;
  date: string;
}

interface OyoHotelReviews {
  averageRating: number;
  totalCount: number;
  reviews: OyoHotelReview[];
}

interface OyoHotel {
  id: string;
  name: string;
  description: string;
  address: string;
  city: string;
  state: string;
  country: string;
  zipCode: string;
  latitude: number;
  longitude: number;
  starRating: number;
  images?: OyoHotelImage[];
  facilities?: OyoHotelFacility[];
  rooms?: OyoHotelRoom[];
  reviews?: OyoHotelReviews;
}

@Injectable()
export class OyoService {
  private readonly apiUrl: string = '';
  private readonly apiKey: string = '';
  private readonly partnerId: string = '';

  // Fallback B2B aggregator credentials
  private readonly hotelbedsApiUrl: string = '';
  private readonly hotelbedsApiKey: string = '';
  private readonly hotelbedsApiSecret: string = '';

  constructor(
    private readonly configService: ConfigService,
    private readonly logger: LoggerService,
    private readonly exceptionService: ExceptionService,
  ) {
    const apiUrl = this.configService.get<string>('OYO_API_URL');
    if (apiUrl) this.apiUrl = apiUrl;

    const apiKey = this.configService.get<string>('OYO_API_KEY');
    if (apiKey) this.apiKey = apiKey;

    const partnerId = this.configService.get<string>('OYO_PARTNER_ID');
    if (partnerId) this.partnerId = partnerId;

    const hotelbedsApiUrl = this.configService.get<string>('HOTELBEDS_API_URL');
    if (hotelbedsApiUrl) this.hotelbedsApiUrl = hotelbedsApiUrl;

    const hotelbedsApiKey = this.configService.get<string>('HOTELBEDS_API_KEY');
    if (hotelbedsApiKey) this.hotelbedsApiKey = hotelbedsApiKey;

    const hotelbedsApiSecret = this.configService.get<string>('HOTELBEDS_API_SECRET');
    if (hotelbedsApiSecret) this.hotelbedsApiSecret = hotelbedsApiSecret;

    this.logger.setContext('OyoService');
  }

  /**
   * Search for hotels based on criteria
   */
  async searchHotels(params: {
    location: string;
    checkInDate: Date;
    checkOutDate: Date;
    adults: number;
    children?: number;
    rooms?: number;
    minPrice?: number;
    maxPrice?: number;
    amenities?: string[];
    propertyType?: string;
    minRating?: number;
  }) {
    try {
      // Check if OYO API is configured
      if (!this.apiUrl || !this.apiKey) {
        this.logger.warn('OYO API not configured, falling back to Hotelbeds API');
        return this.searchHotelsHotelbeds(params);
      }

      // Format dates for OYO API
      const checkInFormatted = params.checkInDate.toISOString().split('T')[0];
      const checkOutFormatted = params.checkOutDate.toISOString().split('T')[0];

      // Prepare request parameters
      const requestParams = {
        partner_id: this.partnerId,
        api_key: this.apiKey,
        location: params.location,
        check_in: checkInFormatted,
        check_out: checkOutFormatted,
        adults: params.adults,
        children: params.children || 0,
        rooms: params.rooms || 1,
        min_price: params.minPrice,
        max_price: params.maxPrice,
        amenities: params.amenities?.join(','),
        property_type: params.propertyType,
        min_rating: params.minRating,
      };

      // Make API request
      const response = await axios.get(`${this.apiUrl}/hotels/search`, {
        params: requestParams,
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
      });

      // Process response
      const hotels = response.data.hotels || [];

      return {
        success: true,
        totalResults: hotels.length,
        hotels: hotels.map((hotel: OyoHotel) => ({
          id: hotel.id,
          name: hotel.name,
          description: hotel.description,
          address: hotel.address,
          city: hotel.city,
          state: hotel.state,
          country: hotel.country,
          zipCode: hotel.zipCode,
          latitude: hotel.latitude,
          longitude: hotel.longitude,
          starRating: hotel.starRating,
          images: hotel.images?.map((image: OyoHotelImage) => image.url) || [],
          lowestPrice: Math.min(...(hotel.rooms?.map((room: OyoHotelRoom) => room.price) || [0])),
          currency: hotel.rooms?.[0]?.currency || 'USD',
          rooms: hotel.rooms?.map((room: OyoHotelRoom) => ({
            id: room.id,
            name: room.name,
            description: room.description,
            maxOccupancy: room.maxOccupancy,
            price: room.price,
            currency: room.currency,
            images: room.images?.map((image: OyoHotelImage) => image.url) || [],
            amenities:
              room.facilities?.map((facility: OyoHotelFacility) => facility.description) || [],
          })),
          reviews:
            hotel.reviews?.reviews?.map((review: OyoHotelReview) => ({
              rating: review.rating,
              comment: review.comment,
              author: review.author,
              date: review.date,
            })) || [],
          averageRating: hotel.reviews?.averageRating || 0,
          reviewCount: hotel.reviews?.totalCount || 0,
        })),
      };
    } catch (error) {
      this.logger.error(`Error searching OYO hotels: ${error.message}`, error.stack);
      this.exceptionService.handleException(error);

      // Fallback to Hotelbeds API
      this.logger.warn('Falling back to Hotelbeds API');
      return this.searchHotelsHotelbeds(params);
    }
  }

  /**
   * Alias for searchHotels to match integration service interface
   */
  async searchProperties(params: Record<string, unknown>) {
    return this.searchHotels(params as any);
  }

  /**
   * Fallback method to search hotels using Hotelbeds API
   */
  private async searchHotelsHotelbeds(params: {
    location: string;
    checkInDate: Date;
    checkOutDate: Date;
    adults: number;
    children?: number;
    rooms?: number;
    minPrice?: number;
    maxPrice?: number;
    amenities?: string[];
    propertyType?: string;
    minRating?: number;
  }) {
    try {
      // Check if Hotelbeds API is configured
      if (!this.hotelbedsApiUrl || !this.hotelbedsApiKey || !this.hotelbedsApiSecret) {
        throw new Error('Hotelbeds API not configured');
      }

      // Format dates for Hotelbeds API
      const checkInFormatted = params.checkInDate.toISOString().split('T')[0];
      const checkOutFormatted = params.checkOutDate.toISOString().split('T')[0];

      // Generate signature for Hotelbeds API
      const timestamp = Math.floor(Date.now() / 1000);
      const signature = this.generateHotelbedsSignature(timestamp);

      // Prepare request body
      const requestBody = {
        stay: {
          checkIn: checkInFormatted,
          checkOut: checkOutFormatted,
        },
        occupancies: [
          {
            rooms: params.rooms || 1,
            adults: params.adults,
            children: params.children || 0,
          },
        ],
        destination: {
          code: params.location,
        },
        filter: {
          minRate: params.minPrice,
          maxRate: params.maxPrice,
          minCategory: params.minRating,
        },
      };

      // Make API request
      const response = await axios.post(
        `${this.hotelbedsApiUrl}/hotel-api/1.0/hotels`,
        requestBody,
        {
          headers: {
            'Api-Key': this.hotelbedsApiKey,
            'X-Signature': signature,
            Accept: 'application/json',
            'Content-Type': 'application/json',
          },
        },
      );

      // Process response
      const hotels = response.data.hotels?.hotels || [];
      const rooms = response.data.hotels?.rooms || [];

      return {
        success: true,
        totalResults: hotels.length,
        hotels: hotels.map((hotel: Record<string, any>) => ({
          id: hotel.code,
          name: hotel.name,
          description: hotel.description || 'No description available',
          address: hotel.address?.content || 'No address available',
          city: hotel.city?.content || 'Unknown',
          state: hotel.state?.content || 'Unknown',
          country: hotel.country?.content || 'Unknown',
          zipCode: hotel.postalCode || 'Unknown',
          latitude: hotel.coordinates?.latitude || 0,
          longitude: hotel.coordinates?.longitude || 0,
          starRating: hotel.categoryCode || 0,
          images: hotel.images?.map((image: Record<string, any>) => image.url) || [],
          lowestPrice: Math.min(
            ...rooms
              .filter((room: Record<string, any>) => room.hotelCode === hotel.code)
              .map((room: Record<string, any>) => room.rates[0]?.net || 0),
          ),
          currency:
            rooms.find((room: Record<string, any>) => room.hotelCode === hotel.code)?.rates[0]
              ?.currency || 'USD',
          rooms: rooms
            .filter((room: Record<string, any>) => room.hotelCode === hotel.code)
            .map((room: Record<string, any>) => ({
              id: room.code,
              name: room.name,
              description: room.description || 'No description available',
              maxOccupancy: room.occupancy?.maxAdults || 2,
              price: room.rates[0]?.net || 0,
              currency: room.rates[0]?.currency || 'USD',
              images: room.images?.map((image: Record<string, any>) => image.url) || [],
              amenities:
                room.facilities?.map((facility: Record<string, any>) => facility.description) || [],
            })),
          reviews: [],
          averageRating: hotel.reviews?.rating || 0,
          reviewCount: hotel.reviews?.count || 0,
        })),
      };
    } catch (error) {
      this.logger.error(`Error searching Hotelbeds hotels: ${error.message}`, error.stack);
      this.exceptionService.handleException(error);

      // Return empty result
      return {
        success: false,
        totalResults: 0,
        hotels: [],
        error: 'Failed to search hotels',
      };
    }
  }

  /**
   * Get hotel details by ID
   */
  async getHotelDetails(hotelId: string) {
    try {
      // Check if OYO API is configured
      if (!this.apiUrl || !this.apiKey) {
        this.logger.warn('OYO API not configured, falling back to Hotelbeds API');
        return this.getHotelDetailsHotelbeds(hotelId);
      }

      // Make API request
      const response = await axios.get(`${this.apiUrl}/hotels/${hotelId}`, {
        params: {
          partner_id: this.partnerId,
          api_key: this.apiKey,
        },
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
      });

      // Process response
      const hotel = response.data.hotel;

      if (!hotel) {
        throw new Error(`Hotel with ID ${hotelId} not found`);
      }

      return {
        success: true,
        hotel: {
          id: hotel.id,
          name: hotel.name,
          description: hotel.description,
          address: hotel.address,
          city: hotel.city,
          state: hotel.state,
          country: hotel.country,
          zipCode: hotel.zipCode,
          latitude: hotel.latitude,
          longitude: hotel.longitude,
          starRating: hotel.starRating,
          images: hotel.images?.map((image: OyoHotelImage) => image.url) || [],
          amenities:
            hotel.facilities?.map((facility: OyoHotelFacility) => facility.description) || [],
          rooms:
            hotel.rooms?.map((room: OyoHotelRoom) => ({
              id: room.id,
              name: room.name,
              description: room.description,
              maxOccupancy: room.maxOccupancy,
              price: room.price,
              currency: room.currency,
              images: room.images?.map((image: OyoHotelImage) => image.url) || [],
              amenities:
                room.facilities?.map((facility: OyoHotelFacility) => facility.description) || [],
            })) || [],
          reviews:
            hotel.reviews?.reviews?.map((review: OyoHotelReview) => ({
              rating: review.rating,
              comment: review.comment,
              author: review.author,
              date: review.date,
            })) || [],
          averageRating: hotel.reviews?.averageRating || 0,
          reviewCount: hotel.reviews?.totalCount || 0,
        },
      };
    } catch (error) {
      this.logger.error(`Error getting OYO hotel details: ${error.message}`, error.stack);
      this.exceptionService.handleException(error);

      // Fallback to Hotelbeds API
      this.logger.warn('Falling back to Hotelbeds API');
      return this.getHotelDetailsHotelbeds(hotelId);
    }
  }

  /**
   * Alias for getHotelDetails to match integration service interface
   */
  async getPropertyDetails(propertyId: string) {
    return this.getHotelDetails(propertyId);
  }

  /**
   * Create a booking
   */
  async createBooking(_bookingData: Record<string, unknown>): Promise<{
    success: boolean;
    booking?: Record<string, unknown>;
    error?: string;
  }> {
    try {
      // Implementation would go here
      return {
        success: true,
        booking: {
          id: 'mock-booking-id',
          // Other booking details
        },
      };
    } catch (error) {
      this.logger.error(`Error creating booking: ${error.message}`, error.stack);
      this.exceptionService.handleException(error);
      return {
        success: false,
        error: 'Failed to create booking',
      };
    }
  }

  /**
   * Fallback method to get hotel details using Hotelbeds API
   */
  private async getHotelDetailsHotelbeds(hotelId: string) {
    try {
      // Check if Hotelbeds API is configured
      if (!this.hotelbedsApiUrl || !this.hotelbedsApiKey || !this.hotelbedsApiSecret) {
        throw new Error('Hotelbeds API not configured');
      }

      // Generate signature for Hotelbeds API
      const timestamp = Math.floor(Date.now() / 1000);
      const signature = this.generateHotelbedsSignature(timestamp);

      // Make API request
      const response = await axios.get(`${this.hotelbedsApiUrl}/hotel-api/1.0/hotels/${hotelId}`, {
        headers: {
          'Api-Key': this.hotelbedsApiKey,
          'X-Signature': signature,
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
      });

      // Process response
      const hotel = response.data.hotel;
      const rooms = response.data.rooms || [];

      if (!hotel) {
        throw new Error(`Hotel with ID ${hotelId} not found`);
      }

      return {
        success: true,
        hotel: {
          id: hotel.code,
          name: hotel.name,
          description: hotel.description || 'No description available',
          address: hotel.address?.content || 'No address available',
          city: hotel.city?.content || 'Unknown',
          state: hotel.state?.content || 'Unknown',
          country: hotel.country?.content || 'Unknown',
          zipCode: hotel.postalCode || 'Unknown',
          latitude: hotel.coordinates?.latitude || 0,
          longitude: hotel.coordinates?.longitude || 0,
          starRating: hotel.categoryCode || 0,
          images: hotel.images?.map((image: Record<string, any>) => image.url) || [],
          amenities:
            hotel.facilities?.map((facility: Record<string, any>) => facility.description) || [],
          rooms: rooms.map((room: Record<string, any>) => ({
            id: room.code,
            name: room.name,
            description: room.description || 'No description available',
            maxOccupancy: room.occupancy?.maxAdults || 2,
            price: room.rates[0]?.net || 0,
            currency: room.rates[0]?.currency || 'USD',
            images: room.images?.map((image: Record<string, any>) => image.url) || [],
            amenities:
              room.facilities?.map((facility: Record<string, any>) => facility.description) || [],
          })),
          reviews: [],
          averageRating: hotel.reviews?.rating || 0,
          reviewCount: hotel.reviews?.count || 0,
        },
      };
    } catch (error) {
      this.logger.error(`Error getting Hotelbeds hotel details: ${error.message}`, error.stack);
      this.exceptionService.handleException(error);

      // Return empty result
      return {
        success: false,
        hotel: null,
        error: 'Failed to get hotel details',
      };
    }
  }

  /**
   * Generate signature for Hotelbeds API
   */
  private generateHotelbedsSignature(timestamp: number): string {
    try {
      return crypto
        .createHash('sha256')
        .update(`${this.hotelbedsApiKey}${this.hotelbedsApiSecret}${timestamp}`)
        .digest('hex');
    } catch (error) {
      this.logger.error(`Error generating signature: ${error.message}`, error.stack);
      return '';
    }
  }
}

File: nestery-flutter/lib/services/oyo_service.dart
import 'package:dio/dio.dart';
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:nestery_flutter/utils/constants.dart';
import 'package:nestery_flutter/utils/api_exception.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

class OyoService {
  final Dio _dio = Dio();
  final FlutterSecureStorage _secureStorage = const FlutterSecureStorage();

  // Singleton pattern
  static final OyoService _instance = OyoService._internal();

  factory OyoService() {
    return _instance;
  }

  OyoService._internal() {
    _initializeDio();
  }

  void _initializeDio() {
    _dio.options.baseUrl = 'https://partner-api.oyorooms.com/v2';
    _dio.options.connectTimeout = const Duration(seconds: 10);
    _dio.options.receiveTimeout = const Duration(seconds: 10);
    _dio.options.headers = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };

    // Add interceptors for logging, error handling, etc.
    _dio.interceptors.add(InterceptorsWrapper(
      onRequest: (options, handler) async {
        // Add API key to all requests
        final apiKey = dotenv.env['OYO_API_KEY'] ?? Constants.oyoApiKey;
        options.headers['X-API-KEY'] = apiKey;

        // Add partner ID to all requests
        final partnerId = dotenv.env['OYO_PARTNER_ID'] ?? Constants.oyoPartnerId;
        options.headers['X-PARTNER-ID'] = partnerId;

        return handler.next(options);
      },
      onResponse: (response, handler) {
        return handler.next(response);
      },
      onError: (DioException e, handler) {
        return handler.next(e);
      },
    ));
  }

  // Search for hotels
  Future<Map<String, dynamic>> searchHotels({
    required String location,
    required DateTime checkInDate,
    required DateTime checkOutDate,
    required int adults,
    int? children,
    int? rooms,
    double? minPrice,
    double? maxPrice,
    List<String>? amenities,
    String? propertyType,
    double? minRating,
  }) async {
    try {
      final response = await _dio.get(
        '/hotels/search',
        queryParameters: {
          'city': location,
          'check_in': _formatDate(checkInDate),
          'check_out': _formatDate(checkOutDate),
          'guests': adults + (children ?? 0),
          'rooms': rooms ?? 1,
          'price_min': minPrice,
          'price_max': maxPrice,
          'amenities': amenities?.join(','),
          'category': propertyType,
          'rating_min': minRating,
          'include_details': true,
        },
      );

      return response.data;
    } on DioException catch (e) {
      // Fallback to B2B aggregator if OYO API fails
      if (e.response?.statusCode == 403 || e.response?.statusCode == 401) {
        return _fallbackToAggregator(
          method: 'searchHotels',
          params: {
            'location': location,
            'checkInDate': checkInDate,
            'checkOutDate': checkOutDate,
            'adults': adults,
            'children': children,
            'rooms': rooms,
            'minPrice': minPrice,
            'maxPrice': maxPrice,
            'amenities': amenities,
            'propertyType': propertyType,
            'minRating': minRating,
          },
        );
      }
      throw ApiException.fromDioError(e);
    } catch (e) {
      throw ApiException(message: 'Failed to search hotels: $e', statusCode: 500);
    }
  }

  // Get hotel details
  Future<Map<String, dynamic>> getHotelDetails({
    required String hotelId,
  }) async {
    try {
      final response = await _dio.get(
        '/hotels/$hotelId',
        queryParameters: {
          'include_rooms': true,
          'include_amenities': true,
          'include_photos': true,
          'include_policies': true,
        },
      );

      return response.data;
    } on DioException catch (e) {
      // Fallback to B2B aggregator if OYO API fails
      if (e.response?.statusCode == 403 || e.response?.statusCode == 401) {
        return _fallbackToAggregator(
          method: 'getHotelDetails',
          params: {
            'hotelId': hotelId,
          },
        );
      }
      throw ApiException.fromDioError(e);
    } catch (e) {
      throw ApiException(message: 'Failed to get hotel details: $e', statusCode: 500);
    }
  }

  // Get room availability
  Future<Map<String, dynamic>> getRoomAvailability({
    required String hotelId,
    required DateTime checkInDate,
    required DateTime checkOutDate,
    required int adults,
    int? children,
    int? rooms,
  }) async {
    try {
      final response = await _dio.get(
        '/hotels/$hotelId/availability',
        queryParameters: {
          'check_in': _formatDate(checkInDate),
          'check_out': _formatDate(checkOutDate),
          'guests': adults + (children ?? 0),
          'rooms': rooms ?? 1,
        },
      );

      return response.data;
    } on DioException catch (e) {
      // Fallback to B2B aggregator if OYO API fails
      if (e.response?.statusCode == 403 || e.response?.statusCode == 401) {
        return _fallbackToAggregator(
          method: 'getRoomAvailability',
          params: {
            'hotelId': hotelId,
            'checkInDate': checkInDate,
            'checkOutDate': checkOutDate,
            'adults': adults,
            'children': children,
            'rooms': rooms,
          },
        );
      }
      throw ApiException.fromDioError(e);
    } catch (e) {
      throw ApiException(message: 'Failed to get room availability: $e', statusCode: 500);
    }
  }

  // Create booking
  Future<Map<String, dynamic>> createBooking({
    required String hotelId,
    required String roomId,
    required DateTime checkInDate,
    required DateTime checkOutDate,
    required int adults,
    required String firstName,
    required String lastName,
    required String email,
    String? phone,
    int? children,
    int? rooms,
    String? specialRequests,
  }) async {
    try {
      final response = await _dio.post(
        '/bookings',
        data: {
          'hotel_id': hotelId,
          'room_id': roomId,
          'check_in': _formatDate(checkInDate),
          'check_out': _formatDate(checkOutDate),
          'guests': adults + (children ?? 0),
          'rooms': rooms ?? 1,
          'guest_details': {
            'first_name': firstName,
            'last_name': lastName,
            'email': email,
            'phone': phone,
          },
          'special_requests': specialRequests,
        },
      );

      return response.data;
    } on DioException catch (e) {
      // Fallback to B2B aggregator if OYO API fails
      if (e.response?.statusCode == 403 || e.response?.statusCode == 401) {
        return _fallbackToAggregator(
          method: 'createBooking',
          params: {
            'hotelId': hotelId,
            'roomId': roomId,
            'checkInDate': checkInDate,
            'checkOutDate': checkOutDate,
            'adults': adults,
            'firstName': firstName,
            'lastName': lastName,
            'email': email,
            'phone': phone,
            'children': children,
            'rooms': rooms,
            'specialRequests': specialRequests,
          },
        );
      }
      throw ApiException.fromDioError(e);
    } catch (e) {
      throw ApiException(message: 'Failed to create booking: $e', statusCode: 500);
    }
  }

  // Get booking details
  Future<Map<String, dynamic>> getBookingDetails({
    required String bookingId,
  }) async {
    try {
      final response = await _dio.get(
        '/bookings/$bookingId',
      );

      return response.data;
    } on DioException catch (e) {
      // Fallback to B2B aggregator if OYO API fails
      if (e.response?.statusCode == 403 || e.response?.statusCode == 401) {
        return _fallbackToAggregator(
          method: 'getBookingDetails',
          params: {
            'bookingId': bookingId,
          },
        );
      }
      throw ApiException.fromDioError(e);
    } catch (e) {
      throw ApiException(message: 'Failed to get booking details: $e', statusCode: 500);
    }
  }

  // Cancel booking
  Future<Map<String, dynamic>> cancelBooking({
    required String bookingId,
  }) async {
    try {
      final response = await _dio.delete(
        '/bookings/$bookingId',
      );

      return response.data;
    } on DioException catch (e) {
      // Fallback to B2B aggregator if OYO API fails
      if (e.response?.statusCode == 403 || e.response?.statusCode == 401) {
        return _fallbackToAggregator(
          method: 'cancelBooking',
          params: {
            'bookingId': bookingId,
          },
        );
      }
      throw ApiException.fromDioError(e);
    } catch (e) {
      throw ApiException(message: 'Failed to cancel booking: $e', statusCode: 500);
    }
  }

  // Fallback to B2B aggregator
  Future<Map<String, dynamic>> _fallbackToAggregator({
    required String method,
    required Map<String, dynamic> params,
  }) async {
    // Create a new Dio instance for the aggregator
    final aggregatorDio = Dio();
    aggregatorDio.options.baseUrl = 'https://api.hotelbeds.com/hotel-api/1.0';
    aggregatorDio.options.connectTimeout = const Duration(seconds: 10);
    aggregatorDio.options.receiveTimeout = const Duration(seconds: 10);
    aggregatorDio.options.headers = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };

    // Add API key and secret to all requests
    final apiKey = dotenv.env['HOTELBEDS_API_KEY'] ?? Constants.hotelbedsApiKey;
    final apiSecret = dotenv.env['HOTELBEDS_API_SECRET'] ?? Constants.hotelbedsApiSecret;

    // Generate signature (X-Signature) based on API key, secret, and timestamp
    final timestamp = DateTime.now().millisecondsSinceEpoch.toString();
    final signature = _generateSignature(apiKey, apiSecret, timestamp);

    aggregatorDio.options.headers['Api-Key'] = apiKey;
    aggregatorDio.options.headers['X-Signature'] = signature;

    try {
      // Map OYO method to Hotelbeds method
      switch (method) {
        case 'searchHotels':
          final response = await aggregatorDio.post(
            '/hotels',
            data: {
              'stay': {
                'checkIn': _formatDate(params['checkInDate']),
                'checkOut': _formatDate(params['checkOutDate']),
              },
              'occupancies': [
                {
                  'rooms': params['rooms'] ?? 1,
                  'adults': params['adults'],
                  'children': params['children'] ?? 0,
                }
              ],
              'destination': {
                'code': params['location'],
              },
              'filter': {
                'minRate': params['minPrice'],
                'maxRate': params['maxPrice'],
                'minCategory': params['minRating'] != null ? (params['minRating'] * 2).round() : null,
                'amenities': params['amenities'],
              },
            },
          );
          return _mapHotelbedsToOyoResponse(response.data, 'searchHotels');

        case 'getHotelDetails':
          final response = await aggregatorDio.get(
            '/hotels/${params['hotelId']}',
          );
          return _mapHotelbedsToOyoResponse(response.data, 'getHotelDetails');

        case 'getRoomAvailability':
          final response = await aggregatorDio.post(
            '/hotels',
            data: {
              'stay': {
                'checkIn': _formatDate(params['checkInDate']),
                'checkOut': _formatDate(params['checkOutDate']),
              },
              'occupancies': [
                {
                  'rooms': params['rooms'] ?? 1,
                  'adults': params['adults'],
                  'children': params['children'] ?? 0,
                }
              ],
              'hotels': {
                'hotel': [params['hotelId']],
              },
            },
          );
          return _mapHotelbedsToOyoResponse(response.data, 'getRoomAvailability');

        case 'createBooking':
          final response = await aggregatorDio.post(
            '/bookings',
            data: {
              'holder': {
                'name': params['firstName'],
                'surname': params['lastName'],
                'email': params['email'],
                'phone': params['phone'],
              },
              'rooms': [
                {
                  'rateKey': params['roomId'],
                  'paxes': [
                    {
                      'roomId': 1,
                      'type': 'AD',
                      'name': params['firstName'],
                      'surname': params['lastName'],
                    }
                  ],
                }
              ],
              'clientReference': 'NESTERY-${DateTime.now().millisecondsSinceEpoch}',
              'remark': params['specialRequests'],
            },
          );
          return _mapHotelbedsToOyoResponse(response.data, 'createBooking');

        case 'getBookingDetails':
          final response = await aggregatorDio.get(
            '/bookings/${params['bookingId']}',
          );
          return _mapHotelbedsToOyoResponse(response.data, 'getBookingDetails');

        case 'cancelBooking':
          final response = await aggregatorDio.delete(
            '/bookings/${params['bookingId']}',
          );
          return _mapHotelbedsToOyoResponse(response.data, 'cancelBooking');

        default:
          throw ApiException(message: 'Unsupported method: $method', statusCode: 400);
      }
    } on DioException catch (e) {
      throw ApiException.fromDioError(e);
    } catch (e) {
      throw ApiException(message: 'Failed to fallback to aggregator: $e', statusCode: 500);
    }
  }

  // Map Hotelbeds response to OYO response format
  Map<String, dynamic> _mapHotelbedsToOyoResponse(Map<String, dynamic> data, String method) {
    // This is a simplified mapping, in a real implementation this would be more comprehensive
    switch (method) {
      case 'searchHotels':
        return {
          'status': 'success',
          'hotels': data['hotels']['hotels'].map((hotel) => {
            'id': hotel['code'],
            'name': hotel['name'],
            'description': hotel['description'],
            'address': hotel['address'],
            'city': hotel['city'],
            'country': hotel['country'],
            'rating': hotel['categoryCode'] / 2,
            'price': hotel['minRate'],
            'currency': hotel['currency'],
            'images': hotel['images'],
            'amenities': hotel['facilities'],
          }).toList(),
        };

      case 'getHotelDetails':
        final hotel = data['hotel'];
        return {
          'status': 'success',
          'hotel': {
            'id': hotel['code'],
            'name': hotel['name'],
            'description': hotel['description'],
            'address': hotel['address'],
            'city': hotel['city'],
            'country': hotel['country'],
            'rating': hotel['categoryCode'] / 2,
            'images': hotel['images'],
            'amenities': hotel['facilities'],
            'rooms': hotel['rooms'],
            'policies': hotel['policies'],
          },
        };

      case 'getRoomAvailability':
        return {
          'status': 'success',
          'rooms': data['hotels']['hotels'][0]['rooms'].map((room) => {
            'id': room['code'],
            'name': room['name'],
            'description': room['description'],
            'price': room['rates'][0]['net'],
            'currency': room['rates'][0]['currency'],
            'capacity': room['capacity'],
            'amenities': room['facilities'],
          }).toList(),
        };

      case 'createBooking':
        return {
          'status': 'success',
          'booking': {
            'id': data['booking']['reference'],
            'hotel_id': data['booking']['hotel']['code'],
            'room_id': data['booking']['rooms'][0]['code'],
            'check_in': data['booking']['hotel']['checkIn'],
            'check_out': data['booking']['hotel']['checkOut'],
            'guest_name': '${data['booking']['holder']['name']} ${data['booking']['holder']['surname']}',
            'guest_email': data['booking']['holder']['email'],
            'guest_phone': data['booking']['holder']['phone'],
            'status': 'confirmed',
            'total_price': data['booking']['totalNet'],
            'currency': data['booking']['currency'],
          },
        };

      case 'getBookingDetails':
        return {
          'status': 'success',
          'booking': {
            'id': data['booking']['reference'],
            'hotel_id': data['booking']['hotel']['code'],
            'room_id': data['booking']['rooms'][0]['code'],
            'check_in': data['booking']['hotel']['checkIn'],
            'check_out': data['booking']['hotel']['checkOut'],
            'guest_name': '${data['booking']['holder']['name']} ${data['booking']['holder']['surname']}',
            'guest_email': data['booking']['holder']['email'],
            'guest_phone': data['booking']['holder']['phone'],
            'status': data['booking']['status'],
            'total_price': data['booking']['totalNet'],
            'currency': data['booking']['currency'],
          },
        };

      case 'cancelBooking':
        return {
          'status': 'success',
          'message': 'Booking cancelled successfully',
          'booking_id': data['booking']['reference'],
          'cancellation_fee': data['booking']['cancellationFee'],
          'refund_amount': data['booking']['refundAmount'],
        };

      default:
        return data;
    }
  }

  // Generate signature for Hotelbeds API
  String _generateSignature(String apiKey, String apiSecret, String timestamp) {
    // In a real implementation, this would use a proper hashing algorithm
    // For now, we'll just return a placeholder
    return 'signature-placeholder';
  }

  // Helper method to format date
  String _formatDate(DateTime date) {
    return '${date.year}-${date.month.toString().padLeft(2, '0')}-${date.day.toString().padLeft(2, '0')}';
  }
}
